neural_structured_learning/examples/notebooks/adversarial_cnn_transfer_learning_fashionmnist.ipynb

```diff
@@ -0,0 +1,3081 @@
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "q6EQErUK49mH"
+   },
+   "source": [
+    "# Adversarial Attacks with Fast Gradient Sign Method (FGSM)\n",
+    "\n",
+    "## What is an adversarial example?\n",
+    "\n",
+    "Adversarial examples are specialised inputs created with the purpose of confusing a neural network, resulting in the misclassification of a given input. These notorious inputs are indistinguishable to the human eye, but cause the network to fail to identify the contents of the image. There are several types of such attacks, however, here the focus is on the fast gradient sign method attack, which is a *white box* attack whose goal is to ensure misclassification. A white box attack is where the attacker has complete access to the model being attacked. One of the most famous examples of an adversarial image shown below is taken from the aforementioned paper.\n",
+    "\n",
+    "![Adversarial Example](https://i.imgur.com/FyYq2Q0.png)\n",
+    "\n",
+    "Here, starting with the image of a panda, the attacker adds small perturbations (distortions) to the original image, which results in the model labelling this image as a gibbon, with high confidence. The process of adding these perturbations is explained below.\n",
+    "\n",
+    "## Fast gradient sign method\n",
+    "\n",
+    "The fast gradient sign method works by using the gradients of the neural network to create an adversarial example. For an input image, the method uses the gradients of the loss with respect to the input image to create a new image that maximises the loss. This new image is called the adversarial image. This can be summarised using the following expression:\n",
+    "$$adv\\_x = x + \\epsilon*\\text{sign}(\\nabla_xJ(\\theta, x, y))$$\n",
+    "\n",
+    "where \n",
+    "\n",
+    "*   adv_x : Adversarial image.\n",
+    "*   x : Original input image.\n",
+    "*   y : Original input label.\n",
+    "*   $\\epsilon$ : Multiplier to ensure the perturbations are small.\n",
+    "*   $\\theta$ : Model parameters.\n",
+    "*   $J$ : Loss.\n",
+    "\n",
+    "An intriguing property here, is the fact that the gradients are taken with respect to the input image. This is done because the objective is to create an image that maximises the loss. \n",
+    "\n",
+    "A method to accomplish this is to find how much each pixel in the image contributes to the loss value, and add a perturbation accordingly. \n",
+    "\n",
+    "This works pretty fast because it is easy to find how each input pixel contributes to the loss by using the chain rule and finding the required gradients. \n",
+    "\n",
+    "Hence, the gradients are taken with respect to the image. In addition, since the model is no longer being trained (thus the gradient is not taken with respect to the trainable variables, i.e., the model parameters), and so the model parameters remain constant. The only goal is to fool an already trained model.\n",
```

Comment:
```suggestion
    "The gradients are taken with respect to the input image because the objective is to create an image that maximizes the loss. A method to accomplish this is to find how much each pixel in the image contributes to the loss value, and add a perturbation accordingly. This works pretty fast because it is easy to find how much each input pixel contributes to the loss by using the chain rule and finding the required gradients. Since our goal here is to attack a model that has already been trained, the gradient is not taken with respect to the trainable variables, i.e., the model parameters, which are now frozen.\n",
```

Reply:
Thanks changed it